File: .github/scripts/test-distribution-boundary.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "pathname"
require "set"

ROOT = Pathname.new(File.expand_path("../..", __dir__))
CLASSIFICATION = ROOT.join("docs/architecture/distribution-classification.json")
IGNORED_LOCAL_ENTRIES = Set.new(%w[.git .bundle .ruby-lsp .DS_Store]).freeze

failures = []

begin
  value = JSON.parse(CLASSIFICATION.read(encoding: "UTF-8"))
rescue StandardError => e
  warn "distribution classification could not be read: #{e.class}: #{e.message}"
  exit 1
end

unless value.is_a?(Hash)
  failures << "distribution classification must be a JSON object"
  value = {}
end

classification = value["topLevelClassification"]
expected_categories = %w[distribution maintainer split].to_set
unless classification.is_a?(Hash) && classification.keys.to_set == expected_categories
  failures << "topLevelClassification must contain exactly distribution, split, and maintainer"
  classification = {}
end

classified = []
classification.each do |category, entries|
  unless entries.is_a?(Array) && entries.all? { |entry| entry.is_a?(String) && !entry.empty? }
    failures << "#{category} entries must be non-empty strings"
    next
  end
  failures << "#{category} entries must be sorted" unless entries == entries.sort
  failures << "#{category} entries must be unique" unless entries.uniq == entries
  classified.concat(entries)
end

failures << "top-level entries may not be multiply classified" unless classified.uniq == classified
actual = ROOT.children.filter_map do |path|
  next if IGNORED_LOCAL_ENTRIES.include?(path.basename.to_s)

  path.basename.to_s
end.sort
failures << "top-level classification mismatch: expected #{actual.inspect}, got #{classified.sort.inspect}" unless actual == classified.sort
failures << "template must be the sole distribution top-level entry" unless classification["distribution"] == ["template"]
failures << "distribution manifest must remain maintainer-owned" unless classification.fetch("maintainer", []).include?("distribution-manifest.json")

failures << "schemaVersion must be integer 1" unless value["schemaVersion"] == 1
failures << "targetDistributionRoot must be template" unless value["targetDistributionRoot"] == "template"
failures << "directCopyDestination must be ." unless value["directCopyDestination"] == "."
failures << "contentTransformationAllowed must be false" unless value["contentTransformationAllowed"] == false

roots = value["targetSourceRoots"]
expected_roots = {
  "distribution" => "template",
  "maintainer" => ".",
  "publicationInterface" => "docs/publication-catalog.json"
}
failures << "targetSourceRoots mismatch" unless roots == expected_roots
expected_roots.each_value do |path_text|
  path = Pathname.new(path_text)
  failures << "unsafe target source root #{path_text.inspect}" if path.absolute? || path.each_filename.include?("..") || path.each_filename.any? { |part| part.downcase == ".git" }
end

profile_model = value["profileModel"]
expected_composable = %w[
  asset-driven
  browser-interface
  headless-service
  knowledge-augmented
  mcp-enabled
  packaged-cli
  script-assisted
]
unless profile_model.is_a?(Hash)
  failures << "profileModel must be an object"
else
  failures << "templateMarker must remain template-scaffold" unless profile_model["templateMarker"] == "template-scaffold"
  failures << "instruction-only must remain the sole exclusive profile" unless profile_model["exclusiveProfiles"] == ["instruction-only"]
  failures << "composable profiles changed" unless profile_model["composableProfiles"] == expected_composable
  failures << "composition rule must retain required-contract union" unless profile_model["compositionRule"] == "union-of-required-contracts"
end

rules = value["requiredSeparationRules"]
unless rules.is_a?(Array) && rules.length >= 6 && rules.all? { |rule| rule.is_a?(String) && !rule.empty? }
  failures << "requiredSeparationRules must contain at least six non-empty strings"
else
  combined = rules.join(" ").downcase
  %w[branch\ root concrete\ skill\ root escape\ template publication profile clean-room].each do |term|
    failures << "required separation rules omit #{term.tr('\\', ' ')}" unless combined.include?(term.tr("\\", " "))
  end
end

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

puts "Skill source and distribution boundary tests passed."

File: .github/scripts/test_distribution_boundary.py
#!/usr/bin/env python3
"""Validate the source/distribution classification boundary."""

from __future__ import annotations

import json
import sys
from pathlib import Path, PurePosixPath

ROOT = Path(__file__).resolve().parents[2]
CLASSIFICATION = ROOT / "docs/architecture/distribution-classification.json"
IGNORED_LOCAL_ENTRIES = {".git", ".bundle", ".ruby-lsp", ".DS_Store"}


def main() -> int:
    failures: list[str] = []
    try:
        value = json.loads(CLASSIFICATION.read_text(encoding="utf-8"))
    except Exception as exc:
        print(
            f"distribution classification could not be read: "
            f"{type(exc).__name__}: {exc}",
            file=sys.stderr,
        )
        return 1

    if not isinstance(value, dict):
        failures.append("distribution classification must be a JSON object")
        value = {}

    classification = value.get("topLevelClassification")
    expected_categories = {"distribution", "maintainer", "split"}
    if not isinstance(classification, dict) or set(classification) != expected_categories:
        failures.append(
            "topLevelClassification must contain exactly distribution, split, and maintainer"
        )
        classification = {}

    classified: list[str] = []
    for category, entries in classification.items():
        if not isinstance(entries, list) or not all(
            isinstance(entry, str) and entry for entry in entries
        ):
            failures.append(f"{category} entries must be non-empty strings")
            continue
        if entries != sorted(entries):
            failures.append(f"{category} entries must be sorted")
        if len(entries) != len(set(entries)):
            failures.append(f"{category} entries must be unique")
        classified.extend(entries)

    if len(classified) != len(set(classified)):
        failures.append("top-level entries may not be multiply classified")

    actual = sorted(
        path.name for path in ROOT.iterdir() if path.name not in IGNORED_LOCAL_ENTRIES
    )
    if actual != sorted(classified):
        failures.append(
            f"top-level classification mismatch: expected {actual!r}, "
            f"got {sorted(classified)!r}"
        )
    if classification.get("distribution") != ["template"]:
        failures.append("template must be the sole distribution top-level entry")
    if "distribution-manifest.json" not in classification.get("maintainer", []):
        failures.append("distribution manifest must remain maintainer-owned")

    if type(value.get("schemaVersion")) is not int or value.get("schemaVersion") != 1:
        failures.append("schemaVersion must be integer 1")
    if value.get("targetDistributionRoot") != "template":
        failures.append("targetDistributionRoot must be template")
    if value.get("directCopyDestination") != ".":
        failures.append("directCopyDestination must be .")
    if value.get("contentTransformationAllowed") is not False:
        failures.append("contentTransformationAllowed must be false")

    roots = value.get("targetSourceRoots")
    expected_roots = {
        "distribution": "template",
        "maintainer": ".",
        "publicationInterface": "docs/publication-catalog.json",
    }
    if roots != expected_roots:
        failures.append("targetSourceRoots mismatch")
    for path_text in expected_roots.values():
        path = PurePosixPath(path_text)
        if path.is_absolute() or ".." in path.parts or any(
            part.lower() == ".git" for part in path.parts
        ):
            failures.append(f"unsafe target source root {path_text!r}")

    profile_model = value.get("profileModel")
    expected_composable = [
        "asset-driven",
        "browser-interface",
        "headless-service",
        "knowledge-augmented",
        "mcp-enabled",
        "packaged-cli",
        "script-assisted",
    ]
    if not isinstance(profile_model, dict):
        failures.append("profileModel must be an object")
    else:
        if profile_model.get("templateMarker") != "template-scaffold":
            failures.append("templateMarker must remain template-scaffold")
        if profile_model.get("exclusiveProfiles") != ["instruction-only"]:
            failures.append("instruction-only must remain the sole exclusive profile")
        if profile_model.get("composableProfiles") != expected_composable:
            failures.append("composable profiles changed")
        if profile_model.get("compositionRule") != "union-of-required-contracts":
            failures.append("composition rule must retain required-contract union")

    rules = value.get("requiredSeparationRules")
    if (
        not isinstance(rules, list)
        or len(rules) < 6
        or not all(isinstance(rule, str) and rule for rule in rules)
    ):
        failures.append(
            "requiredSeparationRules must contain at least six non-empty strings"
        )
    else:
        combined = " ".join(rules).lower()
        for term in (
            "branch root",
            "concrete skill root",
            "escape template",
            "publication",
            "profile",
            "clean-room",
        ):
            if term not in combined:
                failures.append(f"required separation rules omit {term}")

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1

    print("Skill source and distribution boundary tests passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
